ipatests/test_webui/test_selinuxusermap.py

```diff
@@ -104,9 +127,245 @@ def test_actions(self):
+        actions = ActionChains(self.driver)
+        actions.send_keys(Keys.ENTER).perform()
+        self.wait_for_request(d=0.5)
+        self.assert_notification(assert_text='1 item(s) deleted')
+        self.assert_record(selinux.PKEY, negative=True)
+        self.close_notifications()
+
+        # cleanup
+        self.delete(selinux.ENTITY, [selinux.DATA2])
+        self.delete(hbac.RULE_ENTITY, [hbac.RULE_DATA])
+
+    @screenshot
+    def test_add_different_rules(self):
+        """
+        Test adding different SELinux usermap rules
+        """
+        self.init_app()
+
+        self.navigate_to_entity('config')
+        old_selinux_order = self.get_field_value('ipaselinuxusermaporder')
+        new_selinux_order = '{}${}${}${}${}'.format(
+            old_selinux_order,
+            selinux.DATA_MLS_RANGE['add'][1][2],
+            selinux.DATA_MCS_RANGE['add'][1][2],
+            selinux.DATA_MCS_COMMAS['add'][1][2],
+            selinux.DATA_MLS_SINGLE_VAL['add'][1][2])
+        self.fill_input('ipaselinuxusermaporder', new_selinux_order)
+        self.facet_button_click('save')
+
+        # test add MLS range rule
+        self.add_record(selinux.ENTITY, selinux.DATA_MLS_RANGE)
+        self.assert_record(selinux.PKEY_MLS_RANGE)
+
+        # test add MCS range rule
+        self.add_record(selinux.ENTITY, selinux.DATA_MCS_RANGE)
+        self.assert_record(selinux.PKEY_MCS_RANGE)
+
+        # test add MCS rule with commas
+        self.add_record(selinux.ENTITY, selinux.DATA_MCS_COMMAS)
+        self.assert_record(selinux.PKEY_MCS_COMMAS)
+
+        # test add MLS single value rule
+        self.add_record(selinux.ENTITY, selinux.DATA_MLS_SINGLE_VAL)
+        self.assert_record(selinux.PKEY_MLS_SINGLE_VAL)
+
+        # restore original SELinux user map order
+        self.navigate_to_entity('config')
+        self.fill_input('ipaselinuxusermaporder', old_selinux_order)
+        self.facet_button_click('save')
+
+        # cleanup
+        self.delete(selinux.ENTITY, [selinux.DATA_MLS_RANGE,
+                                     selinux.DATA_MCS_RANGE,
+                                     selinux.DATA_MCS_COMMAS,
+                                     selinux.DATA_MLS_SINGLE_VAL])
+
+    @screenshot
+    def test_undo_refresh_reset_update_cancel(self):
+        """
+        Test undo/refresh/reset/update/cancel buttons
+        """
+        self.init_app()
+
+        mod_description = (selinux.DATA['mod'][0][2])
+
+        # test selinux usermap undo button
+        self.add_record(selinux.ENTITY, selinux.DATA)
+        self.navigate_to_record(selinux.PKEY)
+        self.fill_fields(selinux.DATA['mod'])
+        self.assert_undo_button('description')
+        desc_undo_btn_e = '.widget [name="description"] .btn.undo'
+        desc_undo_btn = self.find(desc_undo_btn_e, By.CSS_SELECTOR)
+        desc_undo_btn.click()
+        self.assert_undo_button('description', visible=False)
+        self.verify_btn_action(mod_description)
```

Comment:
these 4 'undo' lines look as a perfect candidate for a new ui_driver method. E.g. click_undo(field)

Reply:
New method introduced.